config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -72,27 +73,46 @@ public long writeConfig(ConfigResource configResource, String userId, Value conf
     // for the same resource into the document store
     synchronized (configResourceLocks.getUnchecked(configResource)) {
       long configVersion = getLatestVersion(configResource) + 1;
+      long updateTimestamp = System.currentTimeMillis();
+      long creationTimestamp;
+      ContextSpecificConfig existingConfig = getConfig(configResource, configVersion - 1);
+      if (ConfigServiceUtils.isNull(existingConfig.getConfig())) {
+        creationTimestamp = updateTimestamp;
+      } else {
+        creationTimestamp = existingConfig.getCreationTimestamp();
+      }
       Key key = new ConfigDocumentKey(configResource, configVersion);
       Document document = new ConfigDocument(configResource.getResourceName(),
           configResource.getResourceNamespace(), configResource.getTenantId(),
-          configResource.getContext(), configVersion, userId, config);
+          configResource.getContext(), configVersion, userId, config, creationTimestamp,
+          updateTimestamp);
       collection.upsert(key, document);
       return configVersion;
     }
   }
 
   @Override
-  public Value getConfig(ConfigResource configResource) throws IOException {
-    Value config = emptyValue();
+  public ContextSpecificConfig getConfig(ConfigResource configResource) throws IOException {
+    return getConfig(configResource, getLatestVersion(configResource));
+  }
+
+  private ContextSpecificConfig getConfig(ConfigResource configResource, long configVersion)
+      throws IOException {
+    ContextSpecificConfig config = emptyConfig(configResource.getContext());
     Filter filter = getConfigResourceFilter(configResource)
-        .and(Filter.eq(VERSION_FIELD_NAME, getLatestVersion(configResource)));
+        .and(Filter.eq(VERSION_FIELD_NAME, configVersion));
     Query query = new Query();
     query.setFilter(filter);
     Iterator<Document> documentIterator = collection.search(query);
     if (documentIterator.hasNext()) {
       String documentString = documentIterator.next().toJson();
       ConfigDocument configDocument = ConfigDocument.fromJson(documentString);
-      config = configDocument.getConfig();
+      if (!ConfigServiceUtils.isNull(configDocument.getConfig())) {
+        config = ContextSpecificConfig.newBuilder().setConfig(configDocument.getConfig())
```

Comment:
We could return directly here and return an empty config directly on 117, right? No real need to init at 101.

Reply:
updated it